### PR TITLE
Feature/maniphest edit 2.0

### PR DESCRIPTION
**Pull request by D0wn10ad**

### CODE REVIEW

File: README.md
Comment (Grokzen) [inline README.md:48]: This block looks not good when viewed as a markdown file. Try to make this into a list with 4 items instea

File: README.md
Comment (Grokzen) [inline README.md:51]: Is `echo` as a command generally available in a windows environment?

File: README.md
Comment (Grokzen) [inline README.md:55]: This package would be a great addition into the setup.py file as an additional `extras_require` entry instead of only having it in the documentation

File: phabfive/cli.py
Comment (Grokzen) [inline phabfive/cli.py:138]: So this would change the core behaviour of how docopt works? Could you elaborate why we don't need this one any longer?

File: phabfive/constants.py
Comment (Grokzen) [inline phabfive/constants.py:34]:
```suggestion
# dict for all default(?) priority type
```

This comment line don't make sense, is it a question for us or something else? In general if this is a question between a list or dict, i think it mostly depends on what type of application it is used for. In the long term we are going to regular Enum:s in python so it is not critical that we sort this out now. If a list works, we can settle for it for now and we update it in the future. But if the comments is not needed, please remove it

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:63]: This comment is a general comment in all changes you submited in this PR. We generally avoid variable names less then 3 characters. Specially single letter characters. Variables should be easily understood from it's name
```suggestion
        r = self.phab.project.search(constraints = {"name": ""})
```

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:70]:
```suggestion
            r = self.phab.project.search(constraints = {"name": ""}, after=a)
```

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:68]:
```suggestion
        # Added for loop to handle phab instances with more than 100 project/tags
```

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:117]: Don't mix java and python standards, we use snake-case so this would be `after_cursor` please update all changes variables that this rule would apply to

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:129]:
```suggestion
            if "owner" not in ticket_data:
```

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:142]: Duplicate code here by creating same list twice, could be simplified

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:146]:
```suggestion
            if "subscribers" not in ticket_data:
```

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:156]: No empty line here

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:170]:
```suggestion
            if "priority" not in ticket_data:
```

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:172]:
```suggestion

            if ticket_data["priority"] not in TICKET_PRIORITIES:
```

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:168]: Comments in general only starts with one # and not three

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:206]: Naked exception catching is generally bad, at least it should be Explicit `except Exception` at least, but a more narrow or detailed exception you want to look for should be used

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:200]: This regex looks like some standard regex that might be better suited in constants.py as a stable pattern to look for, i dont really get either what it looks for here so a comment line might be good

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:213]: Double indent here

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:215]:
```suggestion
                # list of parents should not exceed 100 in a perfect world?
```

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:226]: This mixed old style prints and is not using f-strings, convert to f-strings

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:224]:
```suggestion
                    taskid = "T" + str(p["id"])
```

Around operators there should always be spaces, this is done in several places in the PR

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:267]: Unless it is needed for any of the logic it checks, parenthesis should be avoided as they are not required or needed here

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:296]:
```suggestion
                transactions.append({"type": "priority", "value": ticket_data["priority"]},)
```

Line wrapping here is not needed, we abide by ~160 character width in general and unless you have much nested data, it can be on a single line

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:299]: Same here

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:323]:
```suggestion
                    "value": user_name_to_id_mapping[ticket_data["owner"]],
```

**Review by Grokzen — CHANGES_REQUESTED** [review]
In general these changes look very nice, they need a bit polishing and linting mostly. We have not done in-depth testing yet and once we do that we can give a bit more feedback on the feature content itself

File: README.md
Comment (D0wn10ad) [reply README.md:51]: Hi, I believe so: https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/echo

File: phabfive/constants.py
Comment (D0wn10ad) [reply phabfive/constants.py:34]: I had to admit I was lazy. On one hand I wanted to atleast to have the chance to assign priorities but on another hand I am stuck with my work that I just hardcoded someting instead to do a proper implementation of using /api/maniphest.priority.search to extract the priority that different Phabricator instances setup.... thus the (?) mark - maybe I would put a FIXME: instead?

Comment (D0wn10ad) [reply phabfive/constants.py:34]: and I have to say, been 10+ years before I code again (was Java -> picked up python because I needed something at work to extract data from my company's Phrabricator instance) - appreciate your effort in reviewing my code + guilty as charged.

File: README.md
Comment (D0wn10ad) [reply README.md]: Would adding platform specific requirements makes better sense?

e.g.

```
setup(
    ...,
    install_requires=[
        "pip_system_certs;platform_system=='Windows'",
    ],
)
```

File: phabfive/cli.py
Comment (D0wn10ad) [reply phabfive/cli.py:138]: This is more about reducing/minimizing the impact on systems installed with docopt-ng instead of docopt.

Per [When docopt and docopt-ng are in the same venv, docopt takes precidence](https://github.com/jazzband/docopt-ng/issues/22), the maintainer of docopt-ng is considering a PEP 541 to take over docopt and by then, line 138 of cli.py will break due to the breaking change introduced by docopt-ng: 
in docopt, the variable is `docopt(help=boolean) `
https://github.com/jazzband/docopt-ng/blob/master/docopt/__init__.py line 865
while docopt-ng is `docopt(default_help=boolean)`
https://github.com/docopt/docopt/blob/master/docopt.py line 505

I would suggest either
1) remove `help=True` (as is) and accept the side effect of no more showing help message by default
2) use docopt.__version__ to decide which to import additional module, [packaging](https://packaging.python.org/en/latest/)

```
import doctopt
from packaging import version

if (version.parse(docopt.__version__) >= version.parse("0.6.2"): # docopt is installed
        cli_args = docopt(
            base_args,
            options_first=True,
            version=phabfive.__version__,
            help=True,
        )
else: # docopt-ng is installed
        cli_args = docopt(
            base_args,
            options_first=True,
            version=phabfive.__version__,
            default_help=True,
        )
```

Please let me know which approach would be preferred?

File: phabfive/maniphest.py
Comment (D0wn10ad) [reply phabfive/maniphest.py:142]: nice catch - is a bug - line 140 is about the project list while 142 user list.

File: phabfive/maniphest.py
Comment (D0wn10ad) [reply phabfive/maniphest.py:156]: removed

File: phabfive/maniphest.py
Comment (D0wn10ad) [reply phabfive/maniphest.py:226]: is leftover in testing - removed.

File: README.md
Comment (D0wn10ad) [reply README.md]: changed the structure - was not familiar with md - thanks!

File: phabfive/maniphest.py
Comment (D0wn10ad) [reply phabfive/maniphest.py:117]: changed to follow what you(?) did for project.

File: phabfive/maniphest.py
Comment (D0wn10ad) [reply phabfive/maniphest.py:206]: will only catch exception generated by phabricator module only.

File: phabfive/maniphest.py
Comment (D0wn10ad) [reply phabfive/maniphest.py:213]: This is by design i.e. intended to be this way.

File: phabfive/maniphest.py
Comment (D0wn10ad) [reply phabfive/maniphest.py:200]: on reviewing how it is done by you on others e.g. paste and diffusion, instead of re-inventing the wheel I created 

 ```
   def _validate_identifier(self, repo_id):
        return re.match(f"^{MONOGRAMS['maniphest']}$", repo_id)
```

to handle validation and also subsequent removal of T from the whole monogram and converting it into int.

since we are on this, please consider how would you want to `maniphest_app.info(task_id)` - should all be made uniform - passing monogram/str i.e. like how `maniphest_app.add_comment(ticket_identifier)` is implemented with proper validation within the function call  `maniphest_app.info()` instead of stripping the T before passing into  `maniphest_app.info(task_id)` like now in cli.py?

File: README.md
Comment (Grokzen) [reply README.md]: @D0wn10ad Yeah acctually that makes much more sense, then you don't have to add in any special markdown readme entry either as it would install automatically where really needed

File: phabfive/cli.py
Comment (Grokzen) [reply phabfive/cli.py:138]: ohhh neat, did not know anyone had finally forked out docopt and made a version that was more widely used and got traction on it

So in either case, as long as the help message behaves the same for the toplevel command and for the sublevel commands, then either solution is find really. I think we had some special case with how we parse sub-commands, but i think the `extras(...)` call deals with that. Test out that commands do not pass through unintentionally to the sub-parsers if you simply run for example `phabfive` phabfive maniphest` etc and that correct parser instances deals with them

File: phabfive/maniphest.py
Comment (Grokzen) [reply phabfive/maniphest.py:200]: Yeah we have thought something similar our self. What we might go with instead tho is to make internal models for some of these primitives like a ticket and put all logic and validation within them instead so that we can pass around proper maniphest objects and re-use logic and features from other tools like Django that we would take inspiration from with how they can lazy load objects and data from a database or api backend like here. Passing around strings this way is not optimal one bit and it needs to be fully refactored eventually

File: phabfive/maniphest.py
Comment (Grokzen) [reply phabfive/maniphest.py:213]: Double indent can't be done by design? or are we talking about different things maybe?

File: README.md
Comment (D0wn10ad) [reply README.md]: I would still want to have a mention in readme if possible as this is really something so hidden that I had to research for days in order to use my test env to test.... but code will be added like such.

File: phabfive/cli.py
Comment (D0wn10ad) [reply phabfive/cli.py:138]: I have actually ran using both docopt and docopt-ng but not necessary testing exhaustively to be 100% confident that I covered 100%.... will try 'phabfive phabfive user whoam i' nonetheless.

File: phabfive/maniphest.py
Comment (D0wn10ad) [reply phabfive/maniphest.py:200]: in that case... I will do minimal changes for the time being ;-D

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:73]: There is still single variable names here that we do not want in, at minimum variables should be 3 letters or more. Please update this through out the entire PR

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:113]: So this might not need to be done in this PR yet but one idea for this kind of hack that seems to be needed all over in general, we could either do a simple helper function that wraps the cursor logic with a yield and a few more dynamic stuff for what endpoint to search, but then that code could be reused more efficiently.

In the long run i think we might want to go to the more Django Model based solution where things are wrapped within the logic of the models where we would sort this kind of problem in there and the end-user up here would never know. This however is for a future version and not within this PR

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:187]: Remove parenthesis here, they do not belong in if-case

File: phabfive/maniphest.py
Comment (Grokzen) [inline phabfive/maniphest.py:224]: This should be a f-string and not old style string concat

File: phabfive/maniphest.py
Comment (Grokzen) [reply phabfive/maniphest.py:226]: Either you missed to push or forgot to change this, please check this one out again

File: phabfive/maniphest.py
Comment (Grokzen) [reply phabfive/maniphest.py:267]: Same here, might be unpushed changes but parenthesis still remains

**Review by Grokzen — CHANGES_REQUESTED** [review]
@D0wn10ad A additional changes to be included. After they are sorted we are happy with merging this.

One thing that is missing is a few updates to the `CHANGELOG.md` file that would be included in the release after this is merged

**Comment by Grokzen** [comment]
@D0wn10ad Any updates to this? we would like to get this in as we are soon pushing an update to the bulk import feature and would like this included before making next release 0.4.0

**Comment by Grokzen** [comment]
@D0wn10ad We have been developing a new v2.1 of the bulk importer with much more added functionality in #55 and i have blended in your changes from this PR where possible into my PR to merge our two new functionality so it will in the end be able to perform the same things you added and more.

I tried to base the work off your's here but it had diverged to far off from the first batch of changes we made in the other PR and your's was not completed so i only copied what i could over manually.

If you make any further changes or patches, please do them on the other PR:s content as-is now before we merge, or later when it is back in the master branch. This PR will be mentioned in the changelog so it won't be forgotten

**Pull request closed without merging**
